rfcs/00xx-select-from-subscription.md

```diff
@@ -0,0 +1,188 @@
+---
+feature: select_from_subscription
+authors:
+  - "hzxa21"
+start_date: "2024/05/08"
+---
+
+# 
+
+## Motivation
+
+This RFC aims to provide a solution to enable user to run query on a changelog stream from a MV/Table, like https://github.com/risingwavelabs/rfcs/pull/83,
+but with a different syntax. It uses the `subscription` concept instead of a TVF to represent the changelog stream.
+
+Let's first recap on the current usage on `subscription`:
+```SQL
+CREATE TABLE t (pk int PRIMARY KEY, v int);
+
+CREATE SUBSCRIPTION sub FROM t WITH (
+    retention = '1D'
+);
+
+DECLARE cur SUBSCRIPTION CURSOR FOR sub;
+
+FETCH NEXT FROM cur;
+  pk   |   v   | op | rw_timestamp 
+-------+-------+----+--------------
+ 25184 | 25184 |  1 |             
+(1 row)
+
+FETCH NEXT FROM cur;
+ t.pk  |  t.v  | t.op | rw_timestamp  
+-------+-------+------+---------------
+ 18704 | 18704 |    2 | 1715021447040
+(1 row)
+
+...
+
+```
+
+Basically we support `FETCH NEXT FROM <cursor>` on a `subscription` to get the changelog stream with schema `(original schema | op | rw_timestamp)`.
+
+
+This RFC will extend the usage of `subscription` to support `SELECT` query on it as well. The output schema of the SELECT query is `(original schema | op | rw_timestamp)`.
+
+## Design
+### Batch SELECT on subscription
+```SQL
+CREATE TABLE t (pk int PRIMARY KEY, v int);
+
+INSERT INTO t values (1, 1);
+
+CREATE SUBSCRIPTION sub FROM t WITH (
+    retention = '1D'
+);
+
+
+INSERT INTO t VALUES (2, 2);
+
+UPDATE t SET v = 11 WHERE pk = 1;
+
+# op = 1 insert
+# op = 2 delete
+# op = 3 update_insert
+# op = 4 update_delete
+SELECT * FROM sub;
+ t.pk  |  t.v  | t.op | rw_timestamp  
+-------+-------+------+---------------
+ 2     | 2     |    1 | 1715021447030    
+ 1     | 1     |    4 | 1715021447040
+ 1     | 11    |    3 | 1715021447040
+(3 rows)
+
+SELECT t.pk as pk, t.v as v, CASE WHEN t.op = 2 or t.op = 4 THEN TRUE ELSE FALSE END as is_delete FROM sub;
+ pk    |   v   |   is_delete   
+ ------+-------+------------
+ 2     | 2     |    f     
+ 1     | 1     |    t 
+ 1     | 11    |    f 
+(3 rows)
+```
+
+- Similar to the requirement for CURSOR query, `retention` must be specified when creating a subscription, that is eligble for batch select.
+- Unlike CUSOR query, batch select won't see historical snapshot data when the subscription is created. It only sees the data after the subscription is created. Technically the following two queries are equivalent:
+```SQL
+SELECT * FROM sub;
+
+# equivalent to FETCH NEXT from cursor created with since begin() till rw_timestamp == now()
+
+DECLARE cur SUBSCRIPTION CURSOR FOR sub SINCE BEGIN();
+FETCH NEXT FROM cur;
+...
+```
+
+Use case: one-shot inspection on the MV/Table changelog for debugging
+
+
+### Streaming SELECT on subscription
+```SQL
+CREATE TABLE t (pk int PRIMARY KEY, v int);
+
+CREATE SUBSCRIPTION sub FROM t;
```

Comment:
Will we persist the created subscription to catalog?

It seems wired to persist it to catalog. Users were intended to only create the sink, and the subscription created here is only a sugar to express the semantic of change log, which is more like a temporary thing. If we persist it, users will have to first inspect whether there are some created subscription, and then create the subscription, and then create the sink, and if not, users can just create the subscription without any concern, and then create the sink.

If we won't persist it, maybe we can make the subscription created as a per-session temporary object, or express subscription as CTE, like `WITH sub AS SUBSCRIPTION from t create materialized view as select * from sub`

Reply:
I think we have already designed subscription semantically as a persistent object. So for now, it must be visible accross multiple sessions. 

But I agree with you that we can specifically introduce a per-session, "temporary subscription" for this user case to address the inconvenience, like:

```

CREATE TABLE t (pk int PRIMARY KEY, v int);

CREATE TEMPORARY SUBSCRIPTION sub FROM t;


CREATE SINK t_changelog_sink
AS SELECT t.pk as pk, t.v as v, CASE WHEN t.op = 2 or t.op = 4 THEN TRUE ELSE FALSE END as is_delete FROM sub;
```

Reply:
If I understand correctly, we should need an execute to convert the op column of streamchunk into a changelog, so persistence is only a syntactic difference (do users need to create subscriptions)?